source/CModuleSystem.cpp

```diff
@@ -0,0 +1,306 @@
+#include "stdafx.h"
+#include "cleo.h"
+#include "CModuleSystem.h"
+
+#include <filesystem>
+#include <fstream>
+
+using namespace CLEO;
+
+void CModuleSystem::Clear()
+{
+	modules.clear();
+}
+
+const ScriptDataRef CModuleSystem::GetExport(const char* moduleName, const char* exportName)
+{
+	// TODO: get current working dir
+	std::string path = "";
+
+	size_t len = strlen(moduleName);
+	if (path.empty())
+	{
+		path.reserve(len);
+	}
+	else
+	{
+		path.reserve(path.length() + len + 1);
+		path.push_back('/'); // path separator
+	}
+
+	path += moduleName;
+	normalizePath(path);
+
+	auto it = modules.find(path);
+	if (it == modules.end()) // module not loaded yet?
+	{
+		if (!LoadFile(path.c_str()))
+		{
+			return {};
+		}
+
+		// check if available now
+		it = modules.find(path);
+		if (it == modules.end())
+		{
+			return {};
+		}
+	}
+	auto& module = it->second;
+
+	return module.GetExport(exportName);
+}
+
+bool CModuleSystem::LoadFile(const char* path)
+{
+	CModule module;
+
+	std::string normalizedPath(path);
+	normalizePath(normalizedPath);
+
+	if (!module.LoadFromFile(normalizedPath.c_str()))
+	{
+		return false;
+	}
+
+	modules[normalizedPath] = std::move(module); // move to container
+	return true;
+}
+
+bool CModuleSystem::LoadDirectory(const char* path)
+{
+	bool result = true;
+
+	for (auto& it : std::filesystem::recursive_directory_iterator(path))
+	{
+		auto& filePath = it.path();
+		if (filePath.extension() == ".s")
+		{
+			result &= LoadFile(filePath.string().c_str());
+		}
+	}
+
+	return result;
+}
+
+bool CModuleSystem::LoadCleoModules()
+{
+	return LoadDirectory("cleo/cleo_modules"); // TODO: select work dir
```

Comment:
modules should be lazy-loaded. there is no need to find and load if no script is using them.

also not sure about pre-defined `cleo_modules` and `.s`. Modules can be anywhere and with any extension.

Reply:
Yes modules are lazy loaded. I had idea there would be "defaul" cleo modules shipped with CLEO, similar to default plugins. In this case it's better to load them all at start. 
Current implementation allows any module extension, no restrictions here.